### ISSUE TITLE
[RFC] remove $identityMap from AggregateRepository

### ISSUE THREAD

**Issue by sandrokeil** [prolic/awesomeness#3]
I would like to discuss to remove the [caching of aggregates via `identityMap`](https://github.com/prolic/awesomeness/blob/b8919c53fe91d45d5c407f3663c0e1a02ee42244/src/EventSourcing/src/Aggregate/AggregateRepository.php#L33). The main problem with this are long running processes like projections. If a projection have to load the aggregate, the wrong state can be loaded. The user has also to ensure not to run in [memory exhausting](https://github.com/prooph/event-sourcing/blob/master/docs/repositories.md#loading-of-thousands-aggregates). This is error prone too.

Assume that PHP process `PhpA` is a long running projection which have to load the aggregate because it replaces the whole read model entry for this aggregate and not only the data from the event. The PHP process `PhpB` is our write process.

```
PhpB -> event1
PhpA -> event1
PhpB -> event2
PhpA -> event2 (but it gets the state of event1 via repository->get() because of identity map)
```

**Comment by prolic** [comment]
I'm okay with removing, what do you think @codeliner ?

**Comment by fritz-gerneth** [comment]
Can second removal. With command handlers in background jobs an AR effectively becomes bound to a specific process (unless the map is cleared). 
1) Process A loads AR, pushes new version 2
2) Process B loads AR, pushes new version 3
3) Process A has AR in cache, tries to push new version, ConcurrencyException but old version is kept in map
4) Command is retried, fails until it reaches process B

We "solved" this by wrapping an repositories save function to just catch concurrency exceptions, clear the cache, and rethrow the exception for all our repos. If caching is required this should be delegated to a strategy or facade imho.

**Comment by sandrokeil** [comment]
Maybe we should do this in current version too, if accepted.

**Comment by fritz-gerneth** [comment]
At least some flag / option on the repo to disable it would be helpful indeed :) Leave it on by default for BC.

**Comment by prolic** [comment]
@sandrokeil @fritz-gerneth 

> At least some flag / option on the repo to disable it would be helpful indeed :) Leave it on by default for BC.

This can be done immediately, as there is no BC, just submit a PR incl. tests and I will merge and release this change.

**Comment by codeliner** [comment]
yeah, we can remove identity map with next major release. In earlier versions we used it to keep the illusion that an aggregate repository can be an in-memory collection (no $repo->save($ar) call needed), but current repository interface has a `save` method and a unit of work is not required/no longer used.

Additional flag to deactivate identity map in current version sounds good, too.
